### ISSUE TITLE
Create proper pull request to add LibreSignal's WebSocket support to OWS Signal

### ISSUE THREAD

**Issue by vanitasvitae** [LibreSignal/LibreSignal#43]
Title sais it all. Some work was done before in https://github.com/WhisperSystems/Signal-Android/issues/1000 but that needs at least to be polished up.

**Comment by vanitasvitae** [comment]
Link to bountysource: https://www.bountysource.com/issues/35722527-create-proper-pull-request-to-add-libresignal-s-websocket-support-to-ows-signal

**Comment by vanitasvitae** [comment]
To make it clear for people, who did not read #37: The pull request must be of a good quality so it will be accepted by @moxie0. 
https://github.com/LibreSignal/LibreSignal/issues/37#issuecomment-226646872

**Comment by BengaloBongali** [comment]
I'm really happy to read this. 2 questions:
1. Will LibreSignal work until WebSocket support was added to Signal?
2. How will "WebSocket-Signal" be distributed (without Play)?

**Comment by vanitasvitae** [comment]
1) If the PR is there in the near future :)
(Libre)Signal has a built-in "time bomb" that warns the user whenever the version is too old. I believe this was the last version of LibreSignal.

2) There is no "WebSocketSignal". There will only be Signal. OWS will not allow distributions outside Google Play until some more points are addressed (eg. crashlogs, user statistics etc)

I have to say, that I'm neither part of the LibreSignal team, nor of OWS. I'm just a user that wants to use Signal without Google Services :)

**Comment by BengaloBongali** [comment]
> (Libre)Signal has a built-in "time bomb" that warns the user whenever the version is too old.

Will it stop working, too?

So the only way to get Signal (with WebSocket) on a phone is and will be https://fdroid.eutopia.cz/?

**Comment by vanitasvitae** [comment]
> Will it stop working, too?

The PR should have no negative impacts on Signals behavior.

> So the only way to get Signal (with WebSocket) on a phone is and will be https://fdroid.eutopia.cz/?

If the PR gets merged, you can use OWS Signal. If you don't have Google Play you can either build it yourselve or use the independent build from https://fdroid.eutopia.cz/

There will be no more need for LibreSignal.

**Comment by BengaloBongali** [comment]
Thanks for your answers!

> The PR should have no negative impacts on Signals behavior.

I meant whether LibreSignal will stop working (before there is WebSocket support in Signal) or is the warning of using an old version all to come? I'm changing phones right now (from one with Google to one without) and I'd like to know whether I can just use LibreSignal now and change to Signal with WebSocket when it's done or I have to think about something in the meantime.

**Comment by vanitasvitae** [comment]
That depends on how quick the PR is ready.

**Comment by bonanza123** [comment]
As there doesn't seem to be much progress with this, we maybe should think about hiring a professional developer. We could try to crowd-source the money, but I guess an issue that we need an estimate of the approx. costs for this. What do you think? Do you have other ideas how to motivate devs to help out here?

**Comment by vanitasvitae** [comment]
I don't think that'll work out. Most Libresignal devs have abandoned the project and switched to other ones. I can fully understand that there is no more motivation to contribute to Signal.

**Comment by bonanza123** [comment]
I didn't mean hiring existing project members, but external that we just hire for this PR. If this websocket problem is solved, the current project members may be more interested again this project.

**Comment by vanitasvitae** [comment]
Of course you can try it, but I doubt that there will be much support. There are a lot of other nice projects that have a much more open attitude towards forks...

**Comment by bonanza123** [comment]
If already you are so sceptically ... ;-) But what those (real) alternatives to Signal?

**Comment by vanitasvitae** [comment]
To be clear, in no way I want to keep you from doing anything :) I'd really appreciate Signal supporting websockets natively. That would solve many issues and if you think you want to do something to achieve that goal, go ahead, I wish you all luck :)

Personally I have switched over to use conversations + OMEMO. Together with Gajim + Gajim-omemo that enables chatting on Linux + Android. The next month or so ChatSecure-iOS will join in providing OMEMO under iOS.

**Comment by mimi89999** [comment]
@bonaza123 Please try Conversations and Gajim with an XMPP server, that supports all Conversations XEPs. They are really great!

I could try to work on that PR, but I doubt it makes much sense...

**Comment by mimi89999** [comment]
The only thing that needs to be added is checking for the presence of GMS core (~GCM client) and if it isn't found, Signal should fail to WebSocket (which is implemented in LibreSignal thanks to @JavaJens)

**Comment by paride** [comment]
What would make sense, in my opinion, is:
- Implementing the WebSocket fallback mechanism;
- Removing the telephony code (RedPhone), as the server is non-free;
- Adding a configuration field where the user can specify the server to use;
- Compiling it with microg gmscore (or removing the GCM stuff altogether).

This would be a generic, "unbranded" client for the Signal server, working with or without gapps (and, if they're present, using the free gmscore library), distribuibile on f-droid, and encouraging the independent deployment of Signal servers and federated networks. It's not easy to setup a Signal server, but it's possible and the procedure is somehow documented.

Removing the RedPhone code (or completely disabling it) is a bold move to show that the generic client is not intended to be used with the official OWS server, but with really free deployments. The user is still free to decide to break the OWS server ToS or not.

**Comment by mimi89999** [comment]
1. Did I say the TLS cert is embeded in the app?
2. Moxie won't like the fact that some users using unofficial clients can use their servers.
3. No federation. It will be very hard for unofficial servers to federate among them self and the official server won't federate. Without federation it doesn't make much sense since people, that you potentially want to contact other than your closest friends probably won't be using the same server...
4. Why to do that since there is XMPP with great clients supporting good encryption: OMEMO

**Comment by paride** [comment]
I didn't know about the TLS cert, and I agree in that federation would be difficult, and impossible with the official server.

It's true that xmpp+omemo works well, but it's also a pity that a really good piece of libre software is "wasted" like this. Another point is that xmpp doesn't allow to register with the phone number and no password, which is what the average user expects nowadays.

**Comment by moxie0** [comment]
> Did I say the TLS cert is embeded in the app?

You can embed your own trust root into the app, or include a CA trust store instead and use CA certs.

> No federation. It will be very hard for unofficial servers to federate among them self and the official server won't federate. Without federation it doesn't make much sense since people, that you potentially want to contact other than your closest friends probably won't be using the same server...

Both the Signal app and the Signal server already support federation, so it should be pretty simple for you to do.

Just because I don't believe this is a viable or desirable strategy doesn't mean that others shouldn't try.  I'd love it if someone proved me wrong, let me know if you need help setting anything up.

**Comment by Heimdal300** [comment]
It would be really nice to have the offical Signal app to support websocket before the last LibreSignal-websocket version runs out. I think apps like conversations or kontalk are not as widespread and advanced as signal. If some starts a crowdfunding I will definitively support it. Please get through it and make the Signal app ready for websocket.

**Comment by vanitasvitae** [comment]
@Heimdal300 you can already go ahead and support the bounty source on this issue ;)

**Comment by Heimdal300** [comment]
@vanitasvitae Done. Thank you for this hint! For others here is the link: https://www.bountysource.com/issues/35722527-create-proper-pull-request-to-add-libresignal-s-websocket-support-to-ows-signal

**Comment by paride** [comment]
If this ever reaches upstream I'll consider setting up an independent Signal server to be used with a rebranded Signal client. The service would allow 3rd party clients and would be fairly open to federation.

What worries me mostly is the cost of the registration SMS messages from Twilio.
This cost needs to be somehow estimated and donations asked accordingly.

While I think I can manage the server stuff, maintaining an Android app is not for me, so this is why I'll seriously consider working on this only if the official client could be used almost unmodified (i.e., gets the websockes patch merged). In this way the only things that would remain to be modified are:
- The branding;
- Disabling RedPhone (easy).

**Comment by ghost** [comment]
Link to discussion on the OWS community forum: https://whispersystems.discoursehosting.net/t/websocket-version-without-gcm-everyone-happy/

**Comment by mimi89999** [comment]
@moxie0

> let me know if you need help setting anything up.

RedPhone server source or doc would help designing GCM-less signaling.

**Comment by ale5000-git** [comment]
Just a question, instead of infinite discussions wouldn't be better to just create a pull request to add WebSocket to main Signal and discuss everything else just after the code is merged?

This discussion just seems too much theoretical.

**Comment by mimi89999** [comment]
@ale5000-git There was already https://github.com/WhisperSystems/Signal-Android/issues/1000 ...

**Comment by ale5000-git** [comment]
But it is an issue, not a pull request.

**Comment by vanitasvitae** [comment]
https://github.com/WhisperSystems/Signal-Android/pull/1960

**Comment by eightbitkid** [comment]
My attempts at a solution to this issue can be found on both [the Android application][1], and [the supporting library][2].

   [1]: https://github.com/WhisperSystems/Signal-Android/pull/5962
   [2]: https://github.com/WhisperSystems/libsignal-service-java/pull/28

**Comment by mimi89999** [comment]
@moxie0 @eightbitkid Just reminding, that I am blocked in the Signal repo, so can't do anything on that...

**Comment by vanitasvitae** [comment]
@mimi89999 Why that? 
@eightbitkid Great work!

**Comment by eightbitkid** [comment]
@vanitasvitae: Thanks!

If anyone has any ideas on how to support voice calls I'd love to hear them. I don't quite see how you receive calls without notifications.

**Comment by mimi89999** [comment]
@eightbitkid @moxie0
I see 3 possibilities:

1. The RedPhone (voice) server implements WebSockets and we register on that server using WebSockets. Of course, then we will have to keep two connections (one to Signal and one to RedPhone). That will ~doubble LibreSignal's/Signal WebSocket battery usage.

2. We do signaling over Signal. That requires even bigger changes to the server.

3. We take the code from microg so that we can talk to GCM servers.

**Comment by anarcat** [comment]
On 2016-12-18 16:33:42, Michel Le Bihan wrote:
> 1. The RedPhone (voice) server implements WebSockets and we register on that server using WebSockets. Of course, then we will have to keep two connections (one to Signal and one to RedPhone). That will ~doubble LibreSignal's/Signal WebSocket battery usage.

that seems like too big of an overhead.

> 2. We do signaling over Signal. That requires even bigger changes to the server.

i would assume this would be the proper way.

both of those require access to the RedPhone server source code, which
is not currently public, AFAIK.

> 3. We take the code from microg so that we can talk to GCM servers.

that is already possible right now - no need to patch signal.

**Comment by mimi89999** [comment]
>> 1. The RedPhone (voice) server implements WebSockets and we register
>on that server using WebSockets. Of course, then we will have to keep
>two connections (one to Signal and one to RedPhone). That will ~doubble
>LibreSignal's/Signal WebSocket battery usage.
>
>that seems like too big of an overhead.
>

I agree. We should avoid that option.

>> 2. We do signaling over Signal. That requires even bigger changes to
>the server.
>
>i would assume this would be the proper way.
>

The preferred, but I don't know if OWS will want to do that...

>both of those require access to the RedPhone server source code, which
>is not currently public, AFAIK.
>
>> 3. We take the code from microg so that we can talk to GCM servers.
>
>that is already possible right now - no need to patch signal.

Right now you need signature spoofing for it to work. Only (very) advanced users know how to do this....

In my suggestion, the code would be directly in Signal, so no need for microg apps and signature spoofing.

**Comment by relyt29** [comment]
I support option 1. I have not seen signal having any particularly bad or aggressive battery usage, and I have been using it for months. Opening two connections instead of one is not such a terrible overhead.

Option 2 is not realistic.

Option 3 is bad. Grabbing MicroG code and embedding it in Signal may allow someone to run free software, but it has the effect of causing the phone to poll google servers for updates. So your software is free instead of nonfree - who cares, you're still phoning home to big brother Google, sending requests to their servers on a regular basis.

**Comment by eightbitkid** [comment]
> The RedPhone (voice) server implements WebSockets and we register on that server using WebSockets.

@mimi89999: Do you have any actual examples of where this might happen in the codebase? Closest I could find was something like [NetworkConnector.java] but the whole API is written to be a 'open, have a conversation of `Signal`s, then close`. Nothing I can find indicates you can leave a connection open.

I stress I don't want to get into some weird hypothetical solutions. I'm just wondering if anyone has anything concrete as a direction to take.

   [1]: https://github.com/eightbitkid/Signal-Android/blob/master/src/org/thoughtcrime/redphone/signaling/NetworkConnector.java

**Comment by mimi89999** [comment]
@eightbitkid As for Doze, I doubt, that Google will have a problem with Signal escaping it since it is only when Play Services are not available.

**Comment by eightbitkid** [comment]
@mimi89999: Siding-stepping battery optimisation is already implemented and no long a concern of mine.

I have asked Moxie if he would accept a patch without voice calling because it appears the source for RedPhone's server is closed. If he doesn't accept then I will be closing the pull request and giving up; it is not possible to work on just the client and hypothetical arguments achieve nothing.

**Comment by mimi89999** [comment]
@eightbitkid Could you ping @moxie0 under that issue to unblock me, please?

**Comment by thestinger** [comment]
@eightbitkid We'll happily host a default-enabled F-Droid repository with this for CopperheadOS whether or not it's accepted, but we need a straightforward way to build and host it. It also needs to be rebranded from Signal to something without Signal in the name. It would be nice if the server was made configurable at runtime too but that's not a requirement for starting out. I would want to keep it as a rebased set of patches on top of the latest stable Signal release, to allow for easy upstreaming it at any point.

If you can help us set that up, I can take care of maintaining it as a clean set of patches indefinitely.

**Comment by paride** [comment]
@thestinger you're surely aware of the fact that moxie opposes third party clients using their servers...

**Comment by thestinger** [comment]
I won't be using their servers, just hosting a modified build of a GPL3 project with their trademark stripped from it. I don't see any legal issues with that. Support for runtime server configuration would be nice, but they choose not to provide that upstream and only supporting their hard-wired server is not a blocker for this to be useful.

**Comment by thestinger** [comment]
Some reading material: https://nplusonemag.com/issue-19/essays/chat-wars/. I'd suggest server-side SafetyNet checks as a technical mechanism to forbid non-stock Android if that's the intent. SafetyNet isn't very good yet, but Google will probably end up turning it into full blown remote attestation over time. Easiest to piggyback on their DRM vs. making it from scratch.

**Comment by thestinger** [comment]
Alright, just need to generate an F-Droid repository.

**Comment by mimi89999** [comment]
@thestinger Maybe XMPP with OMEMO would be better...

**Comment by thestinger** [comment]
CopperheadOS will continue recommending Conversations. It needs an option that's compatible with Signal though.

**Comment by thestinger** [comment]
Reality: people have many contacts with Signal, they have trouble moving them to Conversations. It's by far the number one complaint we get. Nothing else compares. It's the main reason for people wanting Play Services or microG, which isn't happening. So this is a requirement.

The usage guide can be extended to offer this as a fallback option, with caveats explained (drawbacks of being tied to a phone number, poor battery life vs. Conversations at the moment, etc.) and with using Conversations OMEMO whenever possible being highly recommended.

**Comment by mimi89999** [comment]
@thestinger Then I could continue updating LS and rebrand it, if you need that...

**Comment by thestinger** [comment]
I really don't mind maintaining it. I already have it built and in an F-Droid repository. Needed the F-Droid repository anyway, so this works well as a test case. Just need proper production metadata generated for the repository and rebranding.

**Comment by thestinger** [comment]
Also I'd like to keep the changes rebased since it's not really going to be a separate project but rather just maintenance of patches that should land in Signal. I am assuming that they are going to land in Signal, but I'm tired of waiting, so this will be the solution until then.

**Comment by h-2** [comment]
@thestinger I expect more f-droid users will be interested in this. It appears moxie has had some change of heart/mind at least or he wouldn't be considering accepting the PR.  So maybe we shouldn't rush the entire thing right now, and just wait a little?
Since official f-droid now accepts upstream signatures if the builds are reproducible, moxies' main concern about f-droid has been addressed. But like I said, I think maybe we wait for the acceptance patches, and then bring up f-droid again? You could still go for your solution if that fails, and I think we have until end of january until the last LibreSignal build expires...

**Comment by thestinger** [comment]
He said it's not merging the patches as is, so I'm moving forward. I'm not going to assume that he'll be okay with it being redistributed. My assumption is that neither is going to happen. We have to hear about this every day and we need a solution to the problem by the end of the day. I am perfectly fine with it turning into a massive drama. Maintaining it + dealing with drama is a hell of a lot less work than dealing with all of the people asking about Signal and also asking about microG and Play Services primarily because of Signal.

**Comment by h-2** [comment]
> My assumption is that neither is going to happen.

moxie:
> Thanks, needs some work but a good start. This change will also have to include support for calling, [...]

moxie:
> I think the best thing to do is wait for the webrtc_video branch. Shouldn't be too long now, just waiting on iOS.

WebRTC supports seems to be replacing all of redphone so this would make the PR valid. I have often disagreed with Moxie positions in the past, but I think we are closer to a real solution now than we ever were. Starting extra repositories and discussions *now*, might jeopardize this. I understand you are frustrated, but why don't you just use the last LibreSignal build from xmikos repo until we know more?

**Comment by thestinger** [comment]
I need something that can be available by default in CopperheadOS. It needs to be our repository or the default F-Droid repository, and it might as well be our build instead of one from someone else if it's our repository. Even if the changes land, they still need to be shipped in a build that's available in an F-Droid search on CopperheadOS by default. I fully expect that the changes will land within a few months, but then there will be the F-Droid roadblock again. For most users, it still required enabling unknown sources and cannot do unattended upgrades. I am ready to ship the privileged extension for CopperheadOS when the issues with it are resolved (it's integrated and tested, just disabled) but that doesn't help most F-Droid users. Moxie's complaints about statistics, crash dumps, etc. likely still apply to. The pull request landing doesn't solve this for us.

**Comment by thestinger** [comment]
I don't use Signal. I need this because I need to get fewer emails about it. If Open Whisper Systems wants to hire a secretary for me, I'll start caring about their feelings about it.

**Comment by mimi89999** [comment]
@thestinger Let's wait for the WebRTC branch and see... I suspect, that it will be done within January.

@thestinger I don't want a situation where there will be 3 Signal + WebSocket forks...

@h-2 
>Since official f-droid now accepts upstream signatures if the builds are reproducible, moxies' main concern about f-droid has been addressed. But like I said, I think maybe we wait for the acceptance patches, and then bring up f-droid again? You could still go for your solution if that fails, and I think we have until end of january until the last LibreSignal build expires...

Signal still contains the Google GMS application library. That is unacceptable in/for F-Droid. In LibreSignal that was removed (including location support).

**Comment by thestinger** [comment]
Also consider that the main reason https://github.com/mikeperry-tor/mission-improbable includes Google Play is for Signal. It's useful beyond that, but I don't know if @mikeperry-tor would be enabling it by default if it wasn't needed.

> @thestinger I don't want a situation where there will be 3 Signal + WebSocket forks...

CopperheadOS has a fork of dozens of AOSP repositories, F-Droid, F-Droid privileged extension, etc. so for us this is barely a blip on the radar and will greatly reduce our support workload. I'm still going to be telling people to use Conversations either way.

**Comment by thestinger** [comment]
Also, we needed a default-enabled Copperhead F-Droid repository for other reasons. This is a nice way to finally get it implemented and tested. If the changes land, we'll still need to ship an unbranded build like this. If Signal is shipped in F-Droid, we'll still need the repository just not for Signal anymore.

**Comment by moxie0** [comment]
@thestinger The reason that we have the preferences that we do around people not using the Signal service for their own clients is because of *our* workload and what we're capable of managing with two developers. It sounds like you're stressed out, overworked, and generally coming to terms with what it means to develop and support a software product over a moderate period of time. I hope you can appreciate that we're in the same position, and that plenty of your users email us too.

I have never objected to the idea that Signal could support environments without play services or that it could be distributed outside of Play.  However, to achieve those things in a way that is going to be manageable and sustainable for us requires some work, so I would like the people advocating for it to actually do the work. Almost two years after enumerating those tasks, I've been shocked by how reluctant the "free software" community is to write software. Thanks to @eightbitkid for finally writing some code.

**Comment by thestinger** [comment]
I understand why you didn't implement it and that it's a problem to get support requests for something that doesn't always provide timely message delivery or that drains power (not that GCM is perfectly reliable, since it effectively does the same thing, just across apps). I don't understand / agree with a lot of things, but for those issues I do.

Understanding that doesn't mean I'm not going to not immediately take advantage of there finally being an up-to-date, clean set of patches. I need to make sure they stay maintained / tested and the only way to do that is to get a few users testing and to maintain it myself from this point until it lands. Having something that works without invasive/crazy changes breaking the security model / over-the-air updates and without a high technical barrier (building CopperheadOS) also means not having to deal with the endless questions about this anymore and all of the fallout from people using one of the available bad options. One of those bad options is using the abandoned / out-of-date LibreSignal build counting down towards not working anymore. I used to point people in that direction, but had to stop.

I'm preparing for this not landing, the submitter losing interest (I'd have lost interest by now if it was me submitting it), the patches rotting away from unrelated changes, etc. If that happens, someone will just be able to submit our set of patches from a branch tracking the newest Signal version.

It has the warning, already avoids the Signal name or branding (icon) and it could even require explicit configuration for the server choice (needing key fingerprints makes that a bit painful) but I'm not aware of any other Signal servers right now so it's pointless. I wouldn't mind hosting one and setting it as the default but no one is going to end up using it without federation, they'll just follow third party instructions for switching to the official server, which I'm not going to prevent.

Ideally there would be an open standard for push where the OS could communicate the necessary information to apps for their servers to leverage it, but there's nearly zero interest in making real alternatives to Play Services instead of just having insecure hacks to use Google's implementation... so my assumption is that it won't happen until we pay someone to do it, which isn't happening in the near future, and then there's the bigger issues of getting apps and other OSes to adopt it.

I'm fine with doing something unilaterally and making everyone angry. I'd love to see a proper solution happen, but I'm too busy trying to harden Android and upstream things into AOSP to start worrying about Signal, Clang/LLVM, GCC, binutils, Linux, etc. so those just need to be worked around as needed downstream. The last patch one line bug fix I upstreamed into LLVM required meeting Google engineers at a conference and them helping to get it landed... which still took time. I just have no interest in trying to deal with any of that right now. Stuff just has to happen.

**Comment by ale5000-git** [comment]
@thestinger: +1000 :)

I think doing it now is the best way, changes can be ported back to Signal later.

But postponing indefinitely like it has been done just cause loss of interest.

**Comment by eightbitkid** [comment]
Well, this exploded like I wasn't expecting. My bad for not keeping
up to date; figured everyone would be having Christmas off!

For what it's worth, I don't actually like GitHub. I'm only on
here because I wanted to supply these patches to Signal. If
someone needs to contact me for some reason I'm happy to
chat via e-mail (mine is littered around here already) or
I'm on [Ricochet](https://ricochet.im/) if I'm on my PC at
<ricochet:bpin4w5hatsigfi7>. Please don't contact me with requests
or speculation.

 > We'll happily host a default-enabled F-Droid repository with
 > this for CopperheadOS whether or not it's accepted, but we need a
 > straightforward way to build and host it. [...] I would want to keep
 > it as a rebased set of patches on top of the latest stable Signal
 > release, to allow for easy upstreaming it at any point.

@thestinger: This is simply enough and I'm happy to help here; my
current build environment is a little bit hacky but it wouldn't be a
lot of effort to formalise it with submodules and some Groovy. It'd
definitely be a set of patches for a set of patches though: my original
patches to Signal, and another set of patches on top to allow for
`./gradlew build`. I'm not really interesting in _forking_ Signal
but I don't think you are either.

 > I don't want a situation where there will be 3 Signal + WebSocket
 > forks...

@mimi89999: I mean no disrespect but the state of LibreSignal's commit
log is the reason I wrote my patches. `git merge` only goes so far. I
have a hard time understanding why people write non-sensical commit
logs, make __massive__ commits, and can't use branching or rebase.

 > I have never objected to the idea that Signal could support
 > environments without play services or that it could be distributed
 > outside of Play.

@moxie0: From everything I've read the second part of that sentence
doesn't really feel true. But that doesn't actually matter, for what
it's worth I'm actually fine with you not officially releasing outside
of Google Play if you don't see it as useful. I'd expect everyone
here to be capable of compiling, signing, and deploying Signal from
source. Surely if you can install CopperheadOS or CyanogenMod you
have experience with software? If you not you're probably doing
yourself damage.

 > [...] I've been shocked by how reluctant the "free software"
 > community is to write software.

As an irrelevant aside: __this__. I'm already annoyed at everyone
having 'ideas' but not actual code and I've only been at this free
software thing for a month or two.

 > Thanks to @eightbitkid for finally writing some code.

I'm getting Internet Points for this; right?

 > I'm preparing for this not landing, the submitter losing interest
 > (I'd have lost interest by now if it was me submitting it), the
 > patches rotting away from unrelated changes, etc. If that happens,
 > someone will just be able to submit our set of patches from a branch
 > tracking the newest Signal version.

@thestinger: You mentioned somewhere you don't use Signal yourself;
I do. Part of the reason I wrote my patches is because I finally left
Google's Android and wanted to keep all my friends on Signal. I will
continue to rebase even if they're not accepted because I _need_
to. Still, you are obviously free to do whatever you wish with my
changes I put them up because I hoped they'd be useful to someone.

I'll admit I'm a bit annoyed that issues I can't actually do anything
about is keeping my changes out of the codebase; particularly as
nothing on the new branches seems to change the actual registration
system (with a cursory glance) so it's not like my changes couldn't
be accepted and then I then implement voice calls later when the tech
is available. What happens if I'd have tried this months ago? Then
there'd have be no option at all.

But being a bit moody with a maintainer is not how you get patches
accepted, and getting these patches accepted would make my life easier.

**Comment by mimi89999** [comment]
@eightbitkid 
>>Thanks to @eightbitkid for finally writing some code.

>I'm getting Internet Points for this; right?

There is a bounty on the LS issue about making the PR upstream :smile:

Thank you for making the PR.

**Comment by moxie0** [comment]
@thestinger If you want to run your own service for your own product, go for it.  I think it would be less work for you to help make Signal what you want it to be, but that's up to you.

**Comment by GildedHonour** [comment]
what's the state of this project? what about bounty for this task?

**Comment by vanitasvitae** [comment]
@GildedHonour: See the very first comment :) the bounty is at 1000$ at the moment.

**Comment by GildedHonour** [comment]
@vanitasvitae but isn't this project abandoned?

**Comment by thestinger** [comment]
@eightbitkid It's okay, I ended up figuring it out. I would like a cleaner solution than the hack I used for a local library build but it's unimportant, gradle is a PITA. It makes the Android build system look painless... I just need to change the icon (to avoid confusion, regardless of the icon license) and then there needs to be a dead simple UX on first launch prompting for the server / fingerprint. So I don't really have much work to do. I don't plan on publishing the builds anywhere other than a repository for CopperheadOS (it won't be shipped by default).

@moxie0 I'd prefer it if Signal had these changes upstream and was available to CopperheadOS users in some sane way (NOT enabling unknown sources and then somehow downloading it from Play and updating it manually). However, it isn't, and I have to hear about it multiple times on a daily basis, so I have to do something about it. I expect most users aren't going to be interested in anything but the OWS server, but I can try to a) discourage using the app until we don't *need* a custom build, b) remove the hard-wired default and instead include known alternatives and a custom entry field. I am not sure if anyone else is running Signal servers though, particularly federated ones. I only want to host one ourselves as a last resort if no one else is willing to do it.

**Comment by thestinger** [comment]
I'd have been willing to help with getting the change landed but it doesn't look like there's much for me to do now. I was planning on getting around to extracting the LibreSignal changes and cleaning them up to keep them as a rebased changeset for someone to use as a starting point but @eightbitkid already submitted a clean set of patches.

All I can really do beyond that is add the F-Droid privileged extension to CopperheadOS once it's a bit more robust for unattended app upgrades. I have the integration ready but it has rough edges and is currently broken for app uninstalls on 7.1.1. If that's not enough to redistribute official Signal builds, I don't have a solution beyond making a rebranded build. It's not enough for it to be available somewhere if there's no sane way for users to get it.

I do expect the changes to land, but then I expect I'll still end up needing Noise but simply without any changes other than rebranding / making it generic (not just an OWS server client).

**Comment by GildedHonour** [comment]
what's the status of the bounty? isn't this project abandoned?

**Comment by vanitasvitae** [comment]
@GildetHonour THIS project is sort of abandoned (meaning there is no more LibreSignal). But that's not the goal of this issue/the bounty. Have you followed the discussion?

**Comment by GildedHonour** [comment]
No. Is bounty available, can I work on it?

**Comment by ale5000-git** [comment]
The bounty is currently $1000.
It is [here](https://www.bountysource.com/issues/35722527-create-proper-pull-request-to-add-libresignal-s-websocket-support-to-ows-signal).

**Comment by GildedHonour** [comment]
but it was $1800, wasn't it? who stole $800 out of it? :))))

**Comment by vanitasvitae** [comment]
@GildetHonour afaik it never was at 1800$

**Comment by herbsmn** [comment]
For reference, here's the repo for Noise https://github.com/copperhead/Noise

**Comment by herbsmn** [comment]
This issue now has a clock on it. There are only 9 days left before LibreSignal stops working. Are LibreSignal users going to be left out in the cold or is someone going to claim the $1055 bounty? Will moxie accept the pull request in time to get APKs published to make the transition seemless? Shiz is getting real. What is the current status of all of this? Anyone know? Will LibreSignal make a last last last version before the clock hits zero?

**Comment by thestinger** [comment]
It's not going to be accepted before support for WebRTC-based voice calls is implemented and then there will be a need to adapt it to work with that. It's not going to happen in 9 days.

**Comment by herbsmn** [comment]
has it already been submitted?

**Comment by thestinger** [comment]
https://github.com/WhisperSystems/Signal-Android/pull/6081 is the current state of things. I have a change set where it doesn't split isPushRegistered but rather relies on checking in 2 places where isRegistered was used for Play Services. It's problematic right now though because the Play client libraries decide to report a noisy error.

**Comment by thestinger** [comment]
@herbsmn Yes, and it's not going to be accepted right now. It is indefinitely blocked until WebRTC support lands. There is no timeline for that. The only thing that can be done is preparing it as much as possible for after WebRTC support lands.

**Comment by herbsmn** [comment]
would you be willing to make a LibreSignal fork for the LibreSignal Diaspora to use while we wait?

**Comment by thestinger** [comment]
I'm willing to maintain a set of changes to submit upstream into Signal. The set of changes in https://github.com/copperhead/Noise/commits/tmp before the Noise commits is the work-in-progress. It just needs to avoid triggering that internal Play Services client library notification/exception. I'm not sure why the method to check if Play Services is available throws an exception when it's not in some circumstances. It seems to happen when it's outside of an activity. Suggestions welcome. I'd rather not add persistent state like a new preference when it's not really necessary. That's all that really needs to be fixed and then it's only blocked on WebRTC support.

**Comment by herbsmn** [comment]
I use @xmikos' experiemental F-Droid repo. I'm not seeing Noise in regular F-Droid. Does it only show up in Copperhead's version of F-Droid or does Copperhead have an F-Droid repo that I can add in order to get it? I don't use Google or the Play Store.

**Comment by thestinger** [comment]
It's straightforward to build just clone https://github.com/copperhead/libsignal-service-java within the Noise directory and build from one of the branches named after Signal tags. It will be simpler once the library change is merged and then the commit to make it build with a local library can go away.

**Comment by thestinger** [comment]
The other issue was just that it needs to only check for Play being missing, not simply out-of-date, etc. which can be recovered from by the user. That's done in that tmp branch but I haven't made the same adjustment for the release branches yet. I guess I can move that over to the releases before the issue of the Play client libraries being annoying outside of the activities is worked around.

**Comment by herbsmn** [comment]
Any chance you'd be willing to make a personal F-Droid repo like @xmikos did with LibreSignal+Noise so people can get the app via F-Droid? I'm assuming a number of people are going to want to download it in the very near future. I'd personally be willing to send a Bitcoin donation to anyone that is willing to do this.

**Comment by thestinger** [comment]
I'm not interested in developing, promoting and distributing a portable app. Noise is not intended to be a LibreSignal replacement. It's not going to remove the Play Services client libraries, etc. It's too much trouble to make a second portable F-Droid repository.

**Comment by thestinger** [comment]
How about cleaning up the existing minimal set of changes, and then getting that landed in Signal once the WebRTC changes land. Then I can get delete the Noise repository and I won't have to think about this anymore.

**Comment by Be-ing** [comment]
@thestinger if you want this so urgently, maybe you could contribute to the issues that are blocking it in Signal, namely the WebRTC stuff.

**Comment by mimi89999** [comment]
@moxie0 When can we expect WebRTC in Signal?

**Comment by thestinger** [comment]
@Be-ing I have a whole OS to maintain, how about someone else does something for once?

**Comment by thestinger** [comment]
It's not even blocked by Signal Android changes, but rather iOS... I'm not going to buy a Macbook and iPhone to start working on a messenger app that I don't use for an operating system that I don't use, simply to get a small set of downstream changes landed. I care a lot less than you think about what happens with Signal but I try to make a good faith attempt to upstream changes. I'm currently a lot more interested in landing https://android-review.googlesource.com/#/c/322378/ and a bunch of follow-up changes than this kind of thing.

**Comment by vanitasvitae** [comment]
@thestinger Thank you for your efforts. I really appreciate it.

**Comment by Be-ing** [comment]
Totally understandable that you don't want to work on upstream iOS features. So let this be a suggestion to anyone who has the skills and feels this project is important, consider contributing to WebRTC calls in Signal iOS.

**Comment by vvug** [comment]
@eightbitkid You have surely seen that the webrtc branch is landing:

https://github.com/WhisperSystems/Signal-Android/commits/3.29.0

Will you reopen your PR?
That >1100$ bounty is also waiting for you.

**Comment by thestinger** [comment]
See https://github.com/WhisperSystems/libsignal-service-java/pull/34. It needs to land before changes can be made to Signal.

**Comment by eightbitkid** [comment]
@Be-ing: Come on man; don't just rock up into conversation you weren't in and be a dick to someone putting in most of the work. It's just salty.

@vvug: I'm still committed to my changes but I thought I was pretty clear that I'm only prepared to work off mainline work. That isn't unreasonable; Moxie doesn't really want this changes (which is fair enough because it is his project) so he isn't going to keep my modifications in mind when making modifications. When it lands mainline any structural changes become his problem too so I'll be safe.

Not going to pretend I don't want the bounty but if _some of ya'll_ could just read a chapter or two of a book on software engineering, then chill that'd be pretty sweet.

I'm feeling alright about this being done by the end of the month-ish based on previous releases. Until then bask in the heroic effort of @thestinger making my patches so accessible and keeping this ticking over.

**Comment by rainerzufall** [comment]
It looks like it is really going happen!! :tada::tada::tada: moxie just made a change with the title "Support for using Signal without Play Services": https://github.com/WhisperSystems/Signal-Android/commit/1669731329bcc32c84e33035a67a2fc22444c24b

**Comment by herbsmn** [comment]
I'm excited by the title. Anxious to hear people's thoughts on the changes.

**Comment by vanitasvitae** [comment]
Who's code was merged? Was it the code from @eightbitkid?

**Comment by thestinger** [comment]
@moxie0 made a substantially cleaner implementation and it has a boot receiver so it will work a bit better. I'm not sure if it has a backup job scheduler to bring it back to life if it gets killed by OOM but it could. I guess it's based on @eightbitkid0's code to some extent but it's mostly different.

**Comment by thestinger** [comment]
There's also the option of providing a way to enable a foreground service in advanced settings which would stop it from being killed but JobScheduler could at least revive it at 15 minute intervals since that's the minimum time modern Android will allow for periodic jobs (smaller values are clamped to that).

**Comment by Be-ing** [comment]
So would it now be possible to make unofficial builds from the upstream Signal source and run them without Google Play Services to communicate with others using the canonical Google Play Store build? Is there anything else that needs to be changed in the upstream source?

**Comment by paride** [comment]
@Be-ing why would you want unofficial builds, at this point?

**Comment by Be-ing** [comment]
Of course I'd prefer to run official builds, but as @nrizzio commented on [the commit](https://github.com/WhisperSystems/Signal-Android/commit/1669731329bcc32c84e33035a67a2fc22444c24b), there is still work to do before @moxie0 would distribute the canonical upstream APK outside of the Google Play Store.

**Comment by paride** [comment]
@Be-ing just download the official apk from the play store then. Why would you prefer it being built by a third party?

**Comment by vanitasvitae** [comment]
@paride Not everyone (especially a substantial subset of privacy aware people) has access to Google Play. Whats the point in making Signal independent from GCM, if you still need to have Google Play installed to get the app in the first place?

**Comment by paride** [comment]
@vanitasvitae of course. But you don't need the Play Store to download APKs from there. There many other ways, check out for example gplaycli (Debian packaged), gplayweb, Yalp Store on f-droid... You could even set up a private f-droid repository, a cronjob downloading updates, and then update everything using your f-droid client on the phone.

And if you don't trust the OWS apk you have to compile it yourself anyway. Trusting a random third party is not a solution in my opinion.

**Comment by ghost** [comment]
What is Google other than a random third party?

**Comment by paride** [comment]
@d9h02f did you ever check the signature of an APK? Apps from the Play Store are built and signed by the developer, not by Google.

**Comment by anarcat** [comment]
can we just assume there are valid reasons not to use google services and move on?

i  have documented, to the best of my knowledge, the current blockers with publishing independent APKs of Signal in https://github.com/xmikos/fdroiddata/issues/29#issuecomment-281364702

**Comment by ale5000-git** [comment]
@Be-ing: You can get all apk that are in the Play Store at APKPure site.

**Comment by Be-ing** [comment]
I trust a random SaaS service claiming to get APKs from the Play Store as little as I trust a website offering "FREE MP3z!!!!!". I'd trust an unofficial FDroid repo more. Or I could build it myself, but I'd like to know if it will actually be functional without Google Play Services before I spend the trouble learning how to cross compile APKs.

**Comment by xmikos** [comment]
I am providing independent builds of Signal as usual in my main F-Droid repository on https://fdroid.eutopia.cz (built from official source code, only renamed to LibreSignal)

**Comment by Be-ing** [comment]
Great, thanks @xmikos! You may want to revise the description of the app in your repository now to reflect the recent changes.

**Comment by mimi89999** [comment]
What happened to eightbitkid account??!

**Comment by xmikos** [comment]
@Be-ing I will upgrade fdroid.eutopia.cz landing page soon

**Comment by Be-ing** [comment]
I am happy to report that secure text messaging with people using the official Signal build from Google Play Store with @xmikos's build from https://fdroid.eutopia.cz/ works. I have not yet tested WebRTC voice or video calling.

**Comment by vanitasvitae** [comment]
Yeah, how about voice calling? Wasn't that the major reason, why @moxie0 rejected the PR in the first place? It would be really nice, if that issue was gone now :)

**Comment by xmikos** [comment]
@vanitasvitae Voice and video calls should work if both you and other side enable it in settings (see https://whispersystems.org/blog/signal-video-calls-beta/)

**Comment by thestinger** [comment]
Ah it does use a foreground service: https://github.com/WhisperSystems/Signal-Android/commit/1669731329bcc32c84e33035a67a2fc22444c24b#diff-9c046c6a5dc41570be2a81403504b78dR112. So it won't die in the background anymore.

**Comment by h-2** [comment]
> What happened to eightbitkid account??!

:question:

**Comment by mimi89999** [comment]
@h-2 I don't see it anymore.

**Comment by h-2** [comment]
@mimi89999 Yep, I know what you mean, I just wanted to underline, as well, that it is puzzling that the account including all comments apparently vanished. What's going on?

**Comment by vanitasvitae** [comment]
@mimi89999 @h-2 According to mails from bountysource, he stopped, then started and stopped again working on the issue today at 4:08 am central European time.

**Comment by h-2** [comment]
I still think they should get the bounty, as the official commit would probably not have happened without them. *Unless of course they were an alter ego of moxie – that would be hilarious.*

**Comment by ghost** [comment]
@vanitasvitae Perhaps it's time to close this issue, now that [WebSocket support has been added to Signal](https://github.com/WhisperSystems/Signal-Android/commit/1669731329bcc32c84e33035a67a2fc22444c24b)? Signal can now also be downloaded from outside of the Google Play Store: https://signal.org/android/apk/

**Comment by vanitasvitae** [comment]
Probably, but the issue is, who gets the bounty? @eightbitkid has vanished. Should we grant the bounty @moxie0 instead? This is the first bounty I created and I don't really know how to react :/

**Comment by moxie0** [comment]
I wrote the code, but I don't want the bounty. I'd suggest returning it to the people who donated.

**Comment by vanitasvitae** [comment]
I think we'll wait two more days or so and if nobody has a better idea, I'll close this issue as "won't fix", so everyone gets refunded. In that case you can still donate to OWS if you like :)

**Comment by vvug** [comment]
Two ideas on the issue:
* Donate the bounty to the Signal's BitHub fund, which goes to reward merged pull request from independent developers, somehow in the spirit of the bouny;
* Change the target of bounty to fix the "location sharing" function in Signal when it's used without gapps. Relevant issues: https://github.com/WhisperSystems/Signal-Android/issues/6351, https://github.com/WhisperSystems/Signal-Android/issues/5042. Changing the target of the bounty is delicate and could be seen as arbitrary, but I think the desire of the donors was to have a fully functional Signal on their gapps-free devices, and this is the only missing tile.

**Comment by mimi89999** [comment]
Let's close it as a `won't fix`

**Comment by saddy001** [comment]
> Signal can now also be downloaded from outside of the Google Play Store: https://signal.org/android/apk/

My sha256sum is
`46ba0a0e52ee512fbd629dcc1a780b7f9a21eaca2cbc212ed3886e614aa3a914  Downloads/Signal-website-release-3.31.4.apk`
But the website says it should be 
`29:F3:4E:5F:27:F2:11:B4:24:BC:5B:F9:D6:71:62:C0 EA:FB:A2:DA:35:AF:35:C1:64:16:FC:44:62:76:BA:26`
However, I believe the problem is on the server side... ?

**Comment by thestinger** [comment]
It says that's the fingerprint of the *signing certificate* not a specific version of the apk.

**Comment by saddy001** [comment]
Okay, thanks. But how can I check this? Here is how far I've [come](https://stackoverflow.com/questions/11361452/getting-certificate-details-from-an-apk):
`unzip -p Downloads/Signal-website-release-3.31.4.apk META-INF/CERT.RSA |openssl pkcs7 -inform DER -noout -print_certs -text | sha256sum 
b59e8f3016fe1805cc5fb560ad17a8b199d2abe02cd6419965c8f8f9e96d3ffd  -
`

**Comment by paride** [comment]
```
$ keytool -printcert -jarfile Signal\ Private\ Messenger_v3.31.0_apkpure.com.apk 
Signer #1:

Signature:

Owner: CN=Whisper Systems, OU=Research and Development, O=Whisper Systems, L=Pittsburgh, ST=PA, C=US
Issuer: CN=Whisper Systems, OU=Research and Development, O=Whisper Systems, L=Pittsburgh, ST=PA, C=US
Serial number: 4bfbebba
Valid from: Tue May 25 17:24:42 CEST 2010 until: Tue May 16 17:24:42 CEST 2045
Certificate fingerprints:
	 MD5:  D9:0D:B3:64:E3:2F:A3:A7:BD:A4:C2:90:FB:65:E3:10
	 SHA1: 45:98:9D:C9:AD:87:28:C2:AA:9A:82:FA:55:50:3E:34:A8:87:93:74
	 SHA256: 29:F3:4E:5F:27:F2:11:B4:24:BC:5B:F9:D6:71:62:C0:EA:FB:A2:DA:35:AF:35:C1:64:16:FC:44:62:76:BA:26
	 Signature algorithm name: SHA1withRSA
	 Version: 3
```

**Comment by saddy001** [comment]
That's it. Maybe this should be mentioned on the download page.

**Comment by vanitasvitae** [comment]
So, should I close the issue then?

**Comment by vanitasvitae** [comment]
Won't fix.

**Comment by vanitasvitae** [comment]
Hm, did anyone receive their refunds by now? I wrote them an email a few days ago, but received no reply :/

**Comment by vanitasvitae** [comment]
The bounty has been refunded to everyone :)
